include/dr/details/enumerate.hpp

```diff
@@ -0,0 +1,59 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include <dr/details/ranges_shim.hpp>
+
+namespace lib {
+
+namespace internal {
+
+namespace {
+
+template <rng::range R> struct range_size {
+  using type = std::size_t;
+};
+
+template <rng::sized_range R> struct range_size<R> {
+  using type = rng::range_size_t<R>;
+};
+
+template <rng::range R> using range_size_t = typename range_size<R>::type;
+
+} // namespace
+
+template <rng::viewable_range R> auto enumerate(R &&r) {
+  using S = range_size_t<R>;
+  // NOTE: This line only necessary due to bug in range-v3 where views
+  //       have non-weakly-incrementable size types. (Standard mandates
+  //       size type must be weakly incrementable.)
+  using W = std::conditional_t<std::weakly_incrementable<S>, S, std::size_t>;
+  if constexpr (rng::sized_range<R>) {
+    return rng::views::zip(rng::views::iota(W{0}, W{rng::size(r)}),
+                           std::forward<R>(r));
+  } else {
+    return rng::views::zip(rng::views::iota(W{0}), std::forward<R>(r));
+  }
+}
+
+class enumerate_adapter_closure {
+public:
+  enumerate_adapter_closure() {}
+
+  template <rng::viewable_range R> auto operator()(R &&r) const {
+    return enumerate(std::forward<R>(r));
+  }
+
+  template <rng::viewable_range R>
+  friend auto operator|(R &&r, const enumerate_adapter_closure &closure) {
+    return enumerate(std::forward<R>(r));
+  }
+};
+
+inline auto enumerate() { return enumerate_adapter_closure(); }
```

Comment:
I'm currently using this enumerate implementation.  It seems to work reasonably well.  However, there is a version of enumerate shipping in [C++23](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2164r9.pdf), and also one in range-v3 I haven't looked at.  We should eventually switch to those long-term.

Reply:
Indeed there is an enumerate in range-v3. I've already removed one implementation of enumerate from our codebase. I propose to use already implemented stuff by default, unless we know we need something different.

Reply:
There's an enumerate in range-v3, but not in current implementations of the standard library, since it was just added to C++23, so this needs to stay in for now.